File: java/com/thmtvz/s2lox/S2loxFunction.java
package com.thmtvz.s2lox;

import java.util.List;

class S2loxFunction implements S2loxCallable{
    
    private final Stmt.Function declaration;
    private final Environment closure;
    private final boolean isInitializer;

    S2loxFunction(Stmt.Function declaration, Environment closure, boolean isInitializer){
	this.isInitializer = isInitializer;
	this.closure = closure;
	this.declaration = declaration;
    }

    S2loxFunction bind(S2loxInstance instance){
	Environment environment = new Environment(closure);
	environment.define("this", instance);
	return new S2loxFunction(declaration, environment, isInitializer);
    }
    
    @Override
    public Object call(Interpreter interpreter, List<Object> arguments){
	Environment environment = new Environment(closure);
	for(int i = 0; i < declaration.params.size(); ++i){
	    environment.define(declaration.params.get(i).lexeme, arguments.get(i));
	}

	try{
	    interpreter.executeBlock(declaration.body, environment);
	} catch (Return returnValue){
	    if(isInitializer) return closure.getAt(0, "this");
	    return returnValue.value;
	}
	if(isInitializer) return closure.getAt(0, "this");
	return null;
    }

    @Override
    public int arity(){
	return declaration.params.size();
    }

    @Override
    public String toString(){
	return "<fun " + declaration.name.lexeme + ">";
    }
}

File: ts/S2loxFunction.ts
import { FunctionStmt } from "./Stmt.js";
import S2ltype from "./S2ltype.js";
import Environment from "./Environment.js";
import S2loxInstance from "./S2loxInstance.js";
import S2loxCallable from "./S2loxCallable.js";
import Interpreter from "./Interpreter.js";
import Return from "./Return.js";

export default class S2loxFunction implements S2loxCallable{

    constructor(
	private readonly isInitializer: boolean,
	private readonly closure: Environment,
	private readonly declaration: FunctionStmt,
    ) {}

    public bind(instance: S2loxInstance): S2loxFunction{
	let environment = new Environment(this.closure);
	environment.define("this", instance);
	return new S2loxFunction(this.isInitializer, environment,
				 this.declaration);
    }

    public call(interpreter: Interpreter, args: S2ltype[]): S2ltype{
	let environment = new Environment(this.closure);
	for(let i = 0; i < this.declaration.params.length; ++i){
	    environment.define(this.declaration.params[i].lexeme,
			       args[i]);
	}

	try{
	    interpreter.executeBlock(this.declaration.body, environment);
	} catch(returnValue){
	    if(returnValue instanceof Return){
		if(this.isInitializer) return this.closure.getAt(0, "this");
		return returnValue.value;
	    }
	}
	if(this.isInitializer) return this.closure.getAt(0, "this");
	return null;
    }

    public arity(){
	return this.declaration.params.length;
    }
    
    public toString(){
	return `<fun ${this.declaration.name.lexeme}>`;
    }
}
